T7803: Make failover route vrf-aware

hedrok opened the pull request:
## Make failover route vrf-aware
<!--- Provide a general summary of your changes in the Title above -->

* Added vrf to next-hop: `set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2`
* Added vrf and interface to target: `set ... next-hop 10.0.0.2 check target '10.120.0.21' vrf blue`; `set ... next-hop 10.0.0.2 check target '10.120.0.21' interface eth1`

Last two are needed when check VRF is not same as route VRF. In this case icmp can use any/both options, tcp requires `vrf` and arp requires `interface`

Key changes:
1. There is now not one configuration file, but many - each VRF writes it's own configuration file. Daemon updates configuration after round of all checks. It deletes routes that are no longer in configuration.
2. Configuration script is called for each VRF and for default one. It only adds/deletes configuration files and starts daemon only if it is not started yet, stops it only when last file is deleted.

Refactoring changes that can be reverted on first demand by any reviewer:
1. Make `debug` global, add `print_debug`, make it log message to journal also. This makes less lines of code for `if debug:`, makes code more readable.
2. Added named tuples for Route, Nexthop and Target configuration. This is partially to cleanup, required for working with routes as sets to easily find what changed via set operations.
3. `get_ip_command_args` returns most of arguments to `ip` command. Just to avoid copy-paste.
4. Configufation script no longer flushes routes - daemon does it on startup and on exit. If daemon really crashes on stop, `failover` routes will dangle till daemon restart. Just seemed cleaner that routes are fully controlled by daemon.

Added test that checks basic operation, vrf and configuration changes.

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T7803

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result

I've tested manually using GNS3 with topology

```
R-main [VyOS]
R-red [VyOS]
R-blue [VyOS]
R-default [VyOS]
PC1 [VPCS]
PC2 [VPCS]
PC3 [VPCS]

R-main.eth0 [10.0.0.1/24] - [10.0.0.2/24] R-red.eth0
R-main.eth1 [10.120.0.1/24] - [10.120.0.21/24] R-blue.eth0
R-main.eth2 [10.40.0.1/24] - [10.40.0.31/24] R-default.eth0
R-red.eth1 [10.11.0.1/24] - PC1 [10.11.0.110]
R-blue.eth1 [10.21.0.1/24] - PC2 [10.21.0.120]
R-default.eth1 [10.31.0.1/24] - PC3 [10.31.0.130]

R-main
    set system host-name R-main
    
    set interfaces ethernet eth0 address 10.0.0.1/24
    set interfaces ethernet eth1 address 10.120.0.1/24
    set interfaces ethernet eth2 address 10.40.0.1/24

    # VRFs:
    set interfaces ethernet eth0 vrf 'red'
    set interfaces ethernet eth1 vrf 'blue'

    set vrf name red table '2000'
    set vrf name blue table '3000'

R-red
    set system host-name R-red
    set interfaces ethernet eth0 address 10.0.0.2/24
    set interfaces ethernet eth1 address 10.11.0.1/24

PC1
    ip 10.11.0.110/24 10.11.0.1

R-blue
    set system host-name R-blue
    set interfaces ethernet eth0 address 10.120.0.21/24
    set interfaces ethernet eth1 address 10.21.0.1/24

PC2
    ip 10.21.0.120/24 10.21.0.1

R-default
    set system host-name R-default
    set interfaces ethernet eth0 address 10.40.0.31/24
    set interfaces ethernet eth1 address 10.31.0.1/24

PC3
    ip 10.31.0.130/24 10.31.0.1

```

Checks (all on R-main):

```
    # == Default vrf route, default vrf check, disable/enalbe link R-main -- R-default
    # to see route: show ip route

    set protocols failover route 10.31.0.130/32 next-hop 10.40.0.31 check timeout '5'
    set protocols failover route 10.31.0.130/32 next-hop 10.40.0.31 interface 'eth2'
    set protocols failover route 10.31.0.130/32 next-hop 10.40.0.31 metric '10'
    set protocols failover route 10.31.0.130/32 next-hop 10.40.0.31 check target '10.40.0.31'

    #+ Check ping, no route vrf
    set protocols failover route 10.31.0.130/32 next-hop 10.40.0.31 check type 'icmp'

    #+ Check arping, no route vrf
    set protocols failover route 10.31.0.130/32 next-hop 10.40.0.31 check type 'arp'

    #+ Check tcp, no route vrf
    set protocols failover route 10.31.0.130/32 next-hop 10.40.0.31 check type 'tcp'
    set protocols failover route 10.31.0.130/32 next-hop 10.40.0.31 check port 6512


    # == Red route, red check, disable/enable link R-main -- R-red
    # to see route: show ip route vrf red

    set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2 check timeout '5'
    set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2 interface 'eth0'
    set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2 metric '10'
    set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2 check target '10.0.0.2'

    #+ Check ping, same vrf
    set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2 check type 'icmp'

    #+ Check arping, same vrf
    set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2 check type 'arp'

    #+ Check tcp, same vrf
    set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2 check type 'tcp'
    set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2 check port 6512

    # == Red route, blue check, disable/enable link R-main -- R-blue
    # to see route: show ip route vrf red

    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check timeout '5'
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 interface 'eth0'
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 metric '10'

    #+ Check ping, other vrf (via vrf)
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check type 'icmp'
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check target '10.120.0.21' vrf blue

    #+ Check arping, other vrf (via interface)
    delete protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check target
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check type 'arp'
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check target '10.120.0.21' interface eth1

    #+ Check tcp, other vrf (via vrf)
    delete protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check target
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check type 'tcp'
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check target '10.120.0.21' vrf blue
    set vrf name red protocols failover route 10.21.0.110/32 next-hop 10.0.0.2 check port 6512
```

For each check disabled/enabled link and watched whether route disappears/appears. to test TCP I used command

```
while true ; do echo 'Listen...'; netcat -l -p 6512 ; done
```

On R-default, R-red or R-blue.

To see arp/icmp packets watched link capture.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [X] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [X] I have linked this PR to one or more Phabricator Task(s)
- [X] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli)  Added test `/usr/libexec/vyos/tests/smoke/cli/test_protocols_failover.py`
- [X] My commit headlines contain a valid Task id
- [X] My change requires a change to the documentation (https://docs.vyos.io/en/latest/configuration/protocols/failover.html)
- [ ] I have updated the documentation accordingly

hedrok committed:
T7803: Make failover route vrf-aware

* Added 'protocol failover' to 'set vrf name':
  set vrf name red protocols failover route 10.11.0.110/32 next-hop 10.0.0.2
* Added vrf and interface to target:
  set ... next-hop 10.0.0.2 check target '10.120.0.21' vrf blue
  set ... next-hop 10.0.0.2 check target '10.120.0.21' interface eth1

Last two are needed when check VRF is not same as route VRF.
In this case icmp can use any/both options, tcp requires `vrf` and
arp requires `interface`

hedrok commented:
Failed tests:
```
Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_service_dhcp-server.py
Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_service_dhcpv6-server.py
Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_vrf.py
```
All three fail on clean rolling.

sever-sever commented on src/helpers/vyos-failover.py:
```diff
@@ -120,10 +91,17 @@ def is_target_alive(target_list=None,
         match proto:
             case 'icmp':
-                command = f'/usr/bin/ping -q {target} {iface} -n -c 2 -W 1'
+                command = f'/usr/bin/ping -q {target} {iface_opt} -n -c 2 -W 1'
```
Something is wrong here
```
set interfaces ethernet eth0 address '192.168.122.14/24'
set interfaces ethernet eth0 description 'WAN'
set interfaces ethernet eth2 address '192.168.122.145/24'
set interfaces ethernet eth2 description 'WAN-2'
set interfaces ethernet eth2 vrf 'red'
set vrf name red table '1010'

set protocols failover route 1.1.1.1/32 next-hop 192.168.122.1 check target 192.168.122.1 interface 'eth0'
set protocols failover route 1.1.1.1/32 next-hop 192.168.122.1 check target 192.168.122.1 vrf 'red'
set protocols failover route 1.1.1.1/32 next-hop 192.168.122.1 interface 'eth2'
set protocols failover route 1.1.1.1/32 next-hop 192.168.122.1 vrf 'red'

```
With enabled debug
```
sudo systemctl stop vyos-failover.service
sudo /usr/libexec/vyos/vyos-failover.py --config /run/vyos-failover.conf
```
It shows RC0
```
vyos@r14# sudo /usr/libexec/vyos/vyos-failover.py --config /run/vyos-failover.conf
    [ CHECK-TARGET ]: [sudo ip vrf exec red /usr/bin/ping -q 192.168.122.1 -I eth0 -n -c 2 -W 1] -- return-code [RC: 0]
    [ CHECK-TARGET ]: [sudo ip vrf exec red /usr/bin/ping -q 192.168.122.1 -I eth0 -n -c 2 -W 1] -- return-code [RC: 0]
    [ CHECK-TARGET ]: [sudo ip vrf exec red /usr/bin/ping -q 192.168.122.1 -I eth0 -n -c 2 -W 1] -- return-code [RC: 0]
    [ CHECK-TARGET ]: [sudo ip vrf exec red /usr/bin/ping -q 192.168.122.1 -I eth0 -n -c 2 -W 1] -- return-code [RC: 0]
    [ CHECK-TARGET ]: [sudo ip vrf exec red /usr/bin/ping -q 192.168.122.1 -I eth0 -n -c 2 -W 1] -- return-code [RC: 0]
    [ CHECK-TARGET ]: [sudo ip vrf exec red /usr/bin/ping -q 192.168.122.1 -I eth0 -n -c 2 -W 1] -- return-code [RC: 0]
    [ CHECK-TARGET ]: [sudo ip vrf exec red /usr/bin/ping -q 192.168.122.1 -I eth0 -n -c 2 -W 1] -- return-code [RC: 0]

```
Though it is not true:
```
vyos@r14:~$ sudo ip vrf exec red /usr/bin/ping -q 192.168.122.1 -I eth0
PING 192.168.122.1 (192.168.122.1) from 192.168.122.14 eth0: 56(84) bytes of data.
^C
--- 192.168.122.1 ping statistics ---
342 packets transmitted, 342 received, 0% packet loss, time 349195ms
rtt min/avg/max/mdev = 0.135/0.201/0.444/0.038 ms

```
Unexpected route for failed target check present in the system:
```
vyos@r14# run show ip route vrf red kernel 
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, F - PBR,
       f - OpenFabric, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

VRF red:
K>* 1.1.1.1/32 [0/1] via 192.168.122.1, eth2, weight 1, 00:11:45
K * 192.168.122.0/24 [0/0] is directly connected, eth2, weight 1, 00:38:23
[edit]
vyos@r14# 

```

sever-sever replied:
ignore this

sever-sever commented:
It is  impossible to set the default route  with VRF and without VRF
```
set protocols failover route 0.0.0.0/0 next-hop 192.168.122.1 check target 1.1.1.1 vrf 'red'
set protocols failover route 0.0.0.0/0 next-hop 192.168.122.1 vrf 'red'
```
This logic should fix these limits:
```
set protocols failover route 0.0.0.0/0
set vrf name red protocols failover route 0.0.0.0/0
set vrf name blue protocols failover route 0.0.0.0/0
```

hedrok commented:
@sever-sever I've updated PR: added test, refactored a lot (maybe too much), updated interface as discussed.
Please read `Key changes:` and `Refactoring changes` in description. If you don't like any of `Refactoring changes` - just write, I'll revert them.
Thank you.

hedrok commented:
The only test that fails is `test_pppoe_client` - known issue that is already in rolling.
```
2025-10-07T12:30:31.4593625Z DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_interfaces_pppoe.py
2025-10-07T12:30:50.3580168Z DEBUG - test_pppoe_client (__main__.PPPoEInterfaceTest.test_pppoe_client) ... FAIL
```

sever-sever reviewed (approved):
Manual testing works fine
```
set interfaces ethernet eth2 vrf 'red'
set vrf name red table '1010'

set protocols failover route 203.0.113.11/32 next-hop 192.168.122.1 check target 192.168.122.1
set protocols failover route 203.0.113.11/32 next-hop 192.168.122.1 check timeout '1'
set protocols failover route 203.0.113.11/32 next-hop 192.168.122.1 interface 'eth0'
set protocols failover route 203.0.113.11/32 next-hop 192.168.122.1 metric '5'
set protocols failover route 203.0.113.12/32 next-hop 192.168.122.1 check target 192.168.122.1
set protocols failover route 203.0.113.12/32 next-hop 192.168.122.1 check timeout '1'
set protocols failover route 203.0.113.12/32 next-hop 192.168.122.1 interface 'eth0'
set protocols failover route 203.0.113.12/32 next-hop 192.168.122.1 metric '5'

set vrf name red protocols failover route 203.0.113.101/32 next-hop 192.168.122.1 check target 192.168.122.1
set vrf name red protocols failover route 203.0.113.101/32 next-hop 192.168.122.1 check timeout '1'
set vrf name red protocols failover route 203.0.113.101/32 next-hop 192.168.122.1 interface 'eth2'
set vrf name red protocols failover route 203.0.113.102/32 next-hop 192.168.122.1 check target 192.168.122.1
set vrf name red protocols failover route 203.0.113.102/32 next-hop 192.168.122.1 check timeout '1'
set vrf name red protocols failover route 203.0.113.102/32 next-hop 192.168.122.1 check type 'arp'
set vrf name red protocols failover route 203.0.113.102/32 next-hop 192.168.122.1 interface 'eth2'
set vrf name red protocols failover route 203.0.113.103/32 next-hop 192.168.122.1 check target 192.168.122.1
set vrf name red protocols failover route 203.0.113.103/32 next-hop 192.168.122.1 check timeout '1'
set vrf name red protocols failover route 203.0.113.103/32 next-hop 192.168.122.1 check type 'arp'
set vrf name red protocols failover route 203.0.113.103/32 next-hop 192.168.122.1 interface 'eth2'

```
check:
```
vyos@r14# commit
run show i[edit]
vyos@r14# run show ip route kernel 
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, F - PBR,
       f - OpenFabric, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

K>* 203.0.113.11/32 [0/5] via 192.168.122.1, eth0, weight 1, 00:01:42
K>* 203.0.113.12/32 [0/5] via 192.168.122.1, eth0, weight 1, 00:01:39
[edit]
vyos@r14# 
[edit]
vyos@r14# run show ip route vrf red kernel 
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, F - PBR,
       f - OpenFabric, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

VRF red:
K>* 203.0.113.101/32 [0/1] via 192.168.122.1, eth2, weight 1, 00:01:45
K>* 203.0.113.102/32 [0/1] via 192.168.122.1, eth2, weight 1, 00:01:49
K>* 203.0.113.103/32 [0/1] via 192.168.122.1, eth2, weight 1, 00:00:04
[edit]
vyos@r14# 

```

dmbaturin reviewed (approved):
I trust Viacheslav's testing and the change seems safe enough to me.
